### PR TITLE
Feature: skill search form

### PR DESCRIPTION
**Pull request by robinklaassen**
Eerste versie, technisch werkt het maar ik moet nog wat met de CSS spelen om de pagina netter te maken. Maar een eerste review is al welkom als je tijd hebt :)

### CODE REVIEW

File: app/Http/Controllers/MembersController.php [hunk 1]
```diff
@@ -269,4 +269,25 @@ public function search(\Illuminate\Http\Request $request)
...
+	public function searchSkills(\Illuminate\Http\Request $request)
+	{
+		$this->authorize("viewAny", Member::class);
+
+		$skills = $request->input('skills', []);
+		$require_how = $request->input('require_how', 'one');
+
+		$all_skills = Skill::pluck('tag', 'id')->toArray();
+
+		$members = Member::all()->filter(function ($m, $key) use ($skills, $require_how) {
```
Comment (blackshadev) [inline app/Http/Controllers/MembersController.php:283]: I think this can be done more efficient with some sql or eloquent methods. Iterating all members and their skills seem what excessive and doesn't scale well.

File: resources/views/members/search.blade.php [hunk 2]
```diff
@@ -60,8 +60,7 @@
 	<tbody>
 		@forelse ($members as $member)
 		<tr>
-			<td><a href="{{ url('/members', $member->id) }}">{{$member->voornaam}} {{$member->tussenvoegsel}}
-					{{$member->achternaam}}</a></td>
```
Comment (blackshadev) [inline resources/views/members/search.blade.php:64]: Why this change?
Having only the first name isn't that clear since we have 3 Davids. Maybe include the email?

File: resources/views/members/searchSkills.blade.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		<label for="skills">Vaardigheden: </label>
+		<select class="form-control" style="width: 100%;" name="skills[]" id="skills" multiple>
+			@foreach ($all_skills as $id => $skill)
+			<option value="{{$id}}" @if (in_array($id, $skills)) selected @endif>{{$skill}}</option>
+			@endforeach
+		</select>
+    </div>
+    
+    <div class="form-group col-md-4">
+        <label class="radio-inline">
```
Comment (blackshadev) [inline resources/views/members/searchSkills.blade.php:27]: Can you name this "any" that is more descriptive

File: resources/views/members/searchSkills.blade.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+<table class="table table-hover">
+    <thead>
+        <th>Naam</th>
+        <th>Soort lid</th>
+        <th>Vaardigheden & interesses</th>
+    </thead>
+    <tbody>
+        @forelse ($members as $m)
+        <tr>
+            <td><a href="{{ url('/members', $m->id) }}">{{$m->volnaam}}</a></td>
```
Comment (blackshadev) [inline resources/views/members/searchSkills.blade.php:53]: Is only the first name not ambiguous? Since we have 3 david

**Review by blackshadev — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/MembersController.php (hunk 1 above)
Comment (robinklaassen) [reply app/Http/Controllers/MembersController.php:283]: I've looked into using more Eloquent but couldn't really find what I wanted. I can imagine raw SQL working but that's not my strong suit.

File: resources/views/members/search.blade.php (hunk 2 above)
Comment (robinklaassen) [reply resources/views/members/search.blade.php:64]: Changed it to `volnaam` which is the full name...

File: resources/views/members/searchSkills.blade.php (hunk 4 above)
Comment (robinklaassen) [reply resources/views/members/searchSkills.blade.php:53]: `volnaam` not `voornaam` also here

File: resources/views/members/searchSkills.blade.php (hunk 3 above)
Comment (robinklaassen) [reply resources/views/members/searchSkills.blade.php:27]: Yeah good one

File: app/Http/Controllers/MembersController.php (hunk 1 above)
Comment (robinklaassen) [reply app/Http/Controllers/MembersController.php:284]: Hmmm just found something that will help I think: `whereHas` (doc)[https://laravel.com/docs/7.x/eloquent-relationships#querying-relationship-existence]

Comment (blackshadev) [reply app/Http/Controllers/MembersController.php:284]: I will try some things next week. My gut feeling says it is possible. But I fear for bad performance on production with this approach.
It should be possible to get members `with`  skills and filter the skills with where

File: resources/views/members/searchSkills.blade.php (hunk 3 above)
Comment (blackshadev) [reply resources/views/members/searchSkills.blade.php:27]: I don't see it on mobile, but I trust you on it

Comment (blackshadev) [reply resources/views/members/searchSkills.blade.php:27]: I don't see it on mobile, but I trust you on it

File: app/Http/Controllers/MembersController.php (hunk 1 above)
Comment (robinklaassen) [reply app/Http/Controllers/MembersController.php:284]: Trying some now with `whereHas`

Comment (robinklaassen) [reply app/Http/Controllers/MembersController.php]: I think I have a very nice solution for the query now using only Eloquent. I still can't really get the layout to do what I want though 😞

File: app/Http/Controllers/MembersController.php [hunk 5]
```diff
@@ -282,19 +282,10 @@ public function searchSkills(Request $request)
...
-			case 'all':
-				$members = Member::whereHas('skills', function (Builder $query) use ($skills, $require_how) {
-					$query->whereIn('id', $skills);
-				}, '=', count($skills))->get();
-				break;
-		}
+		$num_matches = ($require_how == 'all') ? count($skills) : 1;
+		$members = Member::whereHas('skills', function (Builder $query) use ($skills, $require_how) {
+			$query->whereIn('id', $skills);
+		}, '>=', $num_matches)->get();
```
Comment (blackshadev) [inline app/Http/Controllers/MembersController.php:288]: This is great.
If you need to also have the skills in the view use `with('skills')`

**Review by blackshadev — APPROVED** [review]
Let's go!
